faiss/impl/ProductAdditiveQuantizer.cpp

```diff
@@ -0,0 +1,347 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include <faiss/impl/ProductAdditiveQuantizer.h>
+
+#include <cstddef>
+#include <cstdio>
+#include <cstring>
+#include <memory>
+#include <random>
+
+#include <algorithm>
+
+#include <faiss/impl/AuxIndexStructures.h>
+#include <faiss/impl/FaissAssert.h>
+#include <faiss/utils/distances.h>
+#include <faiss/utils/hamming.h>
+#include <faiss/utils/utils.h>
+
+extern "C" {
+
+// general matrix multiplication
+int sgemm_(
+        const char* transa,
+        const char* transb,
+        FINTEGER* m,
+        FINTEGER* n,
+        FINTEGER* k,
+        const float* alpha,
+        const float* a,
+        FINTEGER* lda,
+        const float* b,
+        FINTEGER* ldb,
+        float* beta,
+        float* c,
+        FINTEGER* ldc);
+}
```

Comment:
By the way, the matrix multiplication code appears in many places. Could we make it a utility function that accepts matrices in row-major order as input? If so, I could do it in the future.

Reply:
In fact I would like to move to the cblas interface (http://www.netlib.org/blas/cblas.h). This would also be good because some platforms don't support the fortran calling conventions (iOS). 
However, it's a lot of tedious work to convert all the calls + make sure it works on all platforms. https://github.com/facebookresearch/faiss/projects/3